assertsprocessor/metrics.go

```diff
@@ -0,0 +1,166 @@
+package assertsprocessor
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/collectors"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+)
+
+type metricHelper struct {
+	logger                *zap.Logger
+	config                *Config
+	prometheusRegistry    *prometheus.Registry
+	latencyHistogram      *prometheus.HistogramVec
+	attributeValueRegExps *map[string]regexp.Regexp
+}
+
+func (p *metricHelper) compileSpanFilterRegexps() error {
+	p.logger.Info("consumer.Start compiling regexps")
+	for attName, matchExpString := range *p.config.AttributeExps {
+		compile, err := regexp.Compile(matchExpString)
+		if err != nil {
+			return err
+		}
+		(*p.attributeValueRegExps)[attName] = *compile
+	}
+	p.logger.Debug("consumer.Start compiled regexps successfully")
+	return nil
+}
+
+// Returns true if a span matches the span selection criteria
+func (p *metricHelper) shouldCaptureMetrics(span ptrace.Span) bool {
+	if len(*p.attributeValueRegExps) > 0 {
+		spanAttributes := span.Attributes()
+		for attName, matchExp := range *p.attributeValueRegExps {
+			value, found := spanAttributes.Get(attName)
+			if !found {
+				return false
+			}
+			//p.logger.Info("Found Span Attribute",
+			//	zap.String(attName, value.AsString()))
+
+			valueMatches := matchExp.String() == value.AsString() || matchExp.MatchString(value.AsString())
+			//p.logger.Info("Value Regexp Result",
+			//	zap.String("regexp", matchExp.String()),
+			//	zap.Bool("result", valueMatches))
+			if !valueMatches {
+				return false
+			}
+		}
+		return true
+	} else {
+		return false
+	}
+}
+
+func (p *metricHelper) captureMetrics(namespace string, service string, span ptrace.Span) {
+	if p.shouldCaptureMetrics(span) {
+		labels := p.buildLabels(namespace, service, span)
+		latencySeconds := computeLatency(span)
+		p.recordLatency(labels, latencySeconds)
+	}
+}
+
+func (p *metricHelper) buildLabels(namespace string, service string, span ptrace.Span) prometheus.Labels {
+	p.logger.Info("consumer.ConsumeTraces Capturing span duration metric for",
```

Comment:
logger.Debug? Presumably, this is for every span?

Reply:
Will switch to Debug in most places. Used Info as it prints by default without having to specify the debug log level at startup.